### PR TITLE
feat: Many Well evaluation, improve prioritized buffer

### PR DESCRIPTION
**Pull request by lollcat**
- Add option for evaluating AIS on the many well using the log weights to estimate the normalising constant.
- Update notebooks accordingly. 
- Improve README
- Improve stability of for replay buffer - by preventing samples with invalid `log_w_adjustment` value from being sampled.
- When sampling multiple batches from the buffer shuffle beforehand so that the distribution across batches is exactly the same (`topk` used in gumbell trick means the batches won't be fully random). 

These changes will effect the results of the paper (it should just improve FAB's performance). Hence I created a separate release for the ICLR paper results and noted this in the README.

### CODE REVIEW

File: fab/utils/prioritised_replay_buffer.py [hunk 1]
```diff
@@ -116,11 +116,17 @@ def adjust(self, log_w_adjustment, log_q, indices):
...
-        self.buffer.log_q_old[indices] = log_q.to(self.device)
+        valid_indices = valid_indices.to(self.device)
+        self.buffer.log_w[valid_indices] += log_w_adjustment.to(self.device)
+        self.buffer.log_q_old[valid_indices] = log_q.to(self.device)
+
+        # Kill samples in the buffer for which the `log_w_adjustment` is invalid.
+        # A common reason this can occur is if AIS discovers a point far outside the reasonable range of the problem.
+        # Which causes nan log probs under the flow.
+        invalid_indices = indices[~valid_adjustment]
+        self.buffer.log_w[invalid_indices] = -torch.ones_like(self.buffer.log_w[invalid_indices])*(float("inf"))
```
Comment (lollcat) [inline fab/utils/prioritised_replay_buffer.py:129]: Prevent invalid samples from being sampled again by setting their importance weight to -inf in the buffer. 
One example of when this can occur is in the Many Well problem, HMC can sometimes return samples way outside the reasonable range that we care about (e.g. 100) causing the flow to return NaN. We remove these samples as they are un-usable for training (cause the loss to be NaN).

File: fab/target_distributions/many_well.py [hunk 2]
```diff
@@ -96,22 +96,23 @@ def log_prob_2D(self, x):
     def performance_metrics(self, samples: torch.Tensor, log_w: torch.Tensor,
                             log_q_fn: Optional[LogProbFunc] = None,
                             batch_size: Optional[int] = None) -> Dict:
-        if log_q_fn is None:
```
Comment (lollcat) [inline fab/target_distributions/many_well.py:99]: If `log_q_fn` is None we can still evaluate the `log_w` accuracy in estimating the normalizing constant.

File: fab/sampling_methods/ais_test.py [hunk 3]
```diff
@@ -12,6 +12,77 @@
 from fab.utils.plotting import plot_history
 from fab.sampling_methods.base import resample
 
+
+def analytic_alpha_2_div(mean_q: torch.Tensor, mean_p: torch.Tensor) -> torch.Tensor:
+    """Calculate \int p(x)^2/q dx where p and q are unit-variance Gaussians."""
+    return torch.exp(torch.sum(mean_p**2 + mean_q**2 - 2*mean_p*mean_q))
+
+
+def test_ais_num_dist(dim: int = 1,
```
Comment (lollcat) [inline fab/sampling_methods/ais_test.py:21]: A nice test that visualises the quality of AIS with increasing number of intermediate distributions.

File: fab/utils/prioritised_replay_buffer.py [hunk 4]
```diff
@@ -11,8 +11,10 @@ def sample_without_replacement(logits: torch.Tensor, n: int) -> torch.Tensor:
     # https://timvieira.github.io/blog/post/2014/07/31/gumbel-max-trick/
     z = torch.distributions.Gumbel(torch.tensor(0.0), torch.tensor(1.0)).sample(
         logits.shape).to(logits.device)
-    topk = torch.topk(z + logits, n)
-    return topk.indices
+    topk = torch.topk(z + logits, n, sorted=False)
+    indices = topk.indices
+    indices = indices[torch.randperm(n).to(indices.device)]
```
Comment (lollcat) [inline fab/utils/prioritised_replay_buffer.py:16]: Ensure indices are randomly ordered

**Review by VincentStimper — APPROVED** [review]
Looks good to me, thanks!
